### PR TITLE
Initial implementation of co-located templates RFC.

### PR DESCRIPTION
**Pull request by rwjblue**
Still quite WIP, but "basically works" when using ember-cli 3.11 and Ember latest canary (including changes from https://github.com/emberjs/ember.js/pull/18158.

### CODE REVIEW

File: ember-addon-main.js [hunk 1]
```diff
@@ -1,8 +1,92 @@
 'use strict';
 
+const fs = require('fs');
 const path = require('path');
 const utils = require('./utils');
 const hashForDep = require('hash-for-dep');
+const walkSync = require('walk-sync');
+const Plugin = require('broccoli-plugin');
+
+class ColocatedTemplateProcessor extends Plugin {
```
Comment (rwjblue) [inline ember-addon-main.js]: FYI - This is absolutely not "ready for prime time", the implementation below is an initial spike (needs help with caching at the least).

**Comment by chriskrycho** [comment]
This is currently breaking with the `@classic` decorator invocation because of how it's transforming the component declaration.

```sh
SyntaxError: /Users/chris/dev/test/new-stuff-yo/new-stuff-yo/components/lame.js: Leading decorators must be attached to a class declaration (6:0)

  4 | 
  5 | @classic
> 6 | const CLASS = class Lame extends Component {
    | ^
  7 | }
  8 | 
  9 | const setComponentTemplate = Ember._setComponentTemplate;
```

This isn't a problem with the defaults, since the default behavior is to generate a `.extend({ ... })`-style class, but for anyone migrating via `@classic` *and* using colocation, it's a breaker, and one without a workaround. (For folks looking on: the transform as written actually matches the ES spec, but the Stage 1 transforms and any TS-dependent editor tooling, including VS Code's, falls down on the revised spec; and all existing docs I'm aware of specify the Stage 1 export position.)

File: package.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "ember-cli-htmlbars",
-  "version": "3.1.0",
+  "version": "3.2.0-alpha.0",
```
Comment (rwjblue) [inline package.json]: We'll need to tweak this, as there are breaking changes now on master.

File: package.json [hunk 3]
```diff
@@ -34,17 +35,22 @@
...
+    "strip-bom": "^3.0.0",
+    "walk-sync": "^1.1.3"
   },
   "devDependencies": {
     "@ember/optional-features": "^0.7.0",
     "broccoli-test-helper": "^2.0.0",
     "chai": "^4.2.0",
     "co": "^4.6.0",
-    "ember-cli": "~3.9.0",
+    "ember-cli": "github:ember-cli/ember-cli",
```
Comment (rwjblue) [inline package.json]: Should be updated to ember-cli@3.12

**Comment by rwjblue** [comment]
@chriskrycho - Can you share a snippet of what you started with? The error doesn't make it clear...

**Comment by chriskrycho** [comment]
Yes, of course! Original input:

```js
import Component from '@ember/component';
import classic from 'ember-classic-decorator';

@classic
export default class Lame extends Component {}
```

**Comment by pzuraq** [comment]
Got an issue popping up over in `ember-set-helper`: https://github.com/pzuraq/ember-set-helper/pull/6

Seems like a custom transform is not being run against colocated templates. I copied the template in the reproduction app into the `app/templates/components` directory and things started working, so it seems to be something with ember-cli and colocation specifically.

**Comment by rwjblue** [comment]
OK, landed the changes from https://github.com/ember-cli/ember-cli-htmlbars/pull/280 into here and that should have addressed quite a few of the issues folks have reported:

* @chriskrycho's comment RE: using decorators before the export
* @pzuraq's comment about preprocessors not running properly (we now do _exactly_ the same as `hbs`, and that works well with custom transforms)

File: ember-addon-main.js (hunk 1 above)
Comment (rwjblue) [reply ember-addon-main.js]: Created https://github.com/ember-cli/ember-cli-htmlbars/issues/290 to track this.

**Pull request merged**